### PR TITLE
Fix subscribed_to maybe empty

### PR DESCRIPTION
**Pull request by uuip**
## Description
According to these pull requests:  https://github.com/celery/celery/pull/7040 and https://github.com/celery/celery/pull/7220, I improved the behavior when subscribed_to has been  empty, no additional requests.

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
could you please add additional unit tests for the changes proposed?

File: celery/backends/redis.py [hunk 1]
```diff
@@ -103,17 +103,22 @@ def _reconnect_pubsub(self):
...
+            for meta in metas:
+                self.on_state_change(self._decode_result(meta), None)
         self._pubsub = self.backend.client.pubsub(
             ignore_subscribe_messages=True,
         )
+        # subscribed_to maybe empty after on_state_change
         if self.subscribed_to:
             self._pubsub.subscribe(*self.subscribed_to)
         else:
-            self._pubsub.ping()
```
Comment (auvipy) [inline celery/backends/redis.py:116]: what are the limitations of self._pubsub.ping() ?

Comment (uuip) [reply celery/backends/redis.py:116]: pubsub will receive an empty reply b'$0\r\n\r\n\0x00...', this has no effect on celery's functionality, but causes extra connections and memory usage.  The reason is not to send a ping, but when redis using hiredis to receive data, each connection has a 64 bytes bytearray. Without sending ping, there are no these empty replies. </br>
 This problem was found when analyzing the high memory after running high-frequency tasks for a long time----there are many bytearray(b'$0\r\n\r\n\0x00...') takes up some memory.  
The unexpected growth in memory is related to redis-py, not celery.
redis-py, I also made some changes [redis-py 3.5.3 less memory usage](https://github.com/uuip/redis-py/tree/3.5)

Comment (auvipy) [reply celery/backends/redis.py:116]: OK can you add additional unit tests for the changes?

Comment (thedrow) [reply celery/backends/redis.py:116]: A unit test that verifies this behavior would be nice to have.

File: celery/backends/redis.py [hunk 2]
```diff
@@ -103,17 +103,22 @@ def _reconnect_pubsub(self):
...
+            self._pubsub.connection = self._pubsub.connection_pool.get_connection(
+                'pubsub', self._pubsub.shard_hint
+            )
+            self._pubsub.connection.register_connect_callback(self._pubsub.on_connect)
```
Comment (thedrow) [inline celery/backends/redis.py:121]: Why was this line added? Can you please add a comment describing what this does?

**Review by thedrow — CHANGES_REQUESTED** [review]
Overall this looks to be the correct fix. I do have a question and we do need a unit test to merge this.

**Comment by uuip** [comment]
@thedrow test_asynchronous.py reduced the coverage, but this pr didn't modify that file.

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
> @thedrow test_asynchronous.py reduced the coverage, but this pr didn't modify that file.

no worries the PR coverage looks good.

**Comment by uuip** [comment]
There is another problem than I haven't submitted.
```python
    def _reconnect_pubsub(self):
        self._pubsub = None
```
```python
    def _reconnect_pubsub(self):
        # release old connection before assign to new connection
        self._pubsub.reset()
```
I'm not sure yet.

**Comment by auvipy** [comment]
> There is another problem than I haven't submitted.
> 
> ```python
>     def _reconnect_pubsub(self):
>         self._pubsub = None
> ```
> 
> ```python
>     def _reconnect_pubsub(self):
>         # release old connection before assign to new connection
>         self._pubsub.reset()
> ```
> 
> I'm not sure yet.

can you please elaborate more?

**Comment by uuip** [comment]
Until 6 hours ago, I completed a test which running 12 hours.   
Logically, the app is very simple, repeating tasks every 10 seconds. 
At the first 5 minute，there were less than 50 connections，at the ending of test, there were actually more than 1200 connections, and their status were not time_wait, but  ESTAB, I suspect that those connections were not closed normally by celery or redis-py. 

> There is another problem than I haven't submitted.
> 
> ```python
>     def _reconnect_pubsub(self):
>         self._pubsub = None
> ```
> 
> ```python
>     def _reconnect_pubsub(self):
>         # release old connection before assign to new connection
>         self._pubsub.reset()
> ```
> 
> I'm not sure yet.

**Comment by auvipy** [comment]
@pawl can you share your thought on last comment of @uuip please?

**Comment by uuip** [comment]
Shall we focus this pr `Fix subscribed_to maybe empty`?  As for the number of connections, I will analyze it again, an issue or pr in future. Thanks for your kind help.

**Comment by auvipy** [comment]
> Shall we focus this pr `Fix subscribed_to maybe empty`? As for the number of connections, I will analyze it again, an issue or pr in future. Thanks for your kind help.

you can also search for related issues & create a new discussion first to have more consensus

**Comment by uuip** [comment]
@thedrow here is a review assigned to you, if you have time.

**Review by auvipy — CHANGES_REQUESTED** [review]
you will need to adjust tests with the codes as well

**Comment by auvipy** [comment]
Also please do more thorough check of codes and tests before another round of review

**Comment by uuip** [comment]
@auvipy I have removed redundant code, please recheck.

**Comment by auvipy** [comment]
ok, i will re check tomorrow. in the mean time can you  please recheck that you are done with the issue you are trying to resolve?

**Comment by uuip** [comment]
When using redis as backend, gevent or eventlet as concurrency, celery app hold too many redis connections. There are some discussion on https://github.com/celery/celery/pull/6985. 
redis-py always use independent connection for every command, one backend instance is enough.
I fix this scene via a workaround but missing spaces around operator -- this cause previous checking error. This is a workaround, not common solution. After removed it, now all checks have passed.

File: celery/backends/redis.py [hunk 3]
```diff
@@ -103,17 +103,25 @@ def _reconnect_pubsub(self):
...
         if self.subscribed_to:
             self._pubsub.subscribe(*self.subscribed_to)
         else:
-            self._pubsub.ping()
+            self._pubsub.connection = self._pubsub.connection_pool.get_connection(
+                'pubsub', self._pubsub.shard_hint
+            )
+            # even if nothing to subscribe, we should not lost the callback after being connected.
+            # maybe on_connect will be changed in the future. at present, this callback
+            # re-subscribe to any channels previously subscribed to.
```
Comment (thedrow) [inline celery/backends/redis.py:123]:
```suggestion
            # even if there is nothing to subscribe, we should not lose the callback after connecting.
            # maybe on_connect will be changed in the future. at present, this callback
            # will re-subscribe to any channels we previously subscribed to.
```

**Review by thedrow — CHANGES_REQUESTED** [review]
Just a minor fix and we're there.

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
thanks both
